File: 138. Copy List with Random Pointer - Medium/solution.cpp
static auto _ = [] () { ios_base::sync_with_stdio(false); cin.tie(nullptr); return 0; } ();

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;

    Node() {}

    Node(int _val, Node* _next, Node* _random) {
        val = _val;
        next = _next;
        random = _random;
    }
};
*/
class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(!head) {
            return nullptr;
        }
        
        if(m[head]) {
            return m[head];
        }
        
        m[head]         = new Node(head->val);
        m[head]->next   = copyRandomList(head->next);
        m[head]->random = copyRandomList(head->random);
        
        return m[head];        
    }

private:
    unordered_map < Node*, Node* > m;
};

File: 138. Copy List with Random Pointer - Medium/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return None

        node = head
        while node:
            copy = Node(node.val, node.next, None)
            node.next = copy
            node = copy.next

        node = head
        while node:
            if node.random:
                node.next.random = node.random.next
            node = node.next.next

        node = head
        copy = copy_head = head.next
        while node:
            node.next = copy.next
            node = node.next

            if node:
                copy.next = node.next
                copy = copy.next

        return copy_head